File: python/ch01_agent/agent_ch01/features/builtin_tools.py
"""第 1 章的 PowerShell 工具。"""

from ..core.commands import CommandRunner
from ..core.tools import ToolDefinition, ToolRegistry, ToolContext, ToolResult, tool_error, tool_success


def create_shell_tool(command_runner: CommandRunner) -> ToolDefinition:
    """构造 shell 工具；handler 只依赖 CommandRunner 接口，便于注入 Fake。

    这个函数类似 Spring `@Bean` 方法：它把工具元数据和真正执行逻辑组装在一起。
    """
    def handler(arguments: dict[str, object], context: ToolContext) -> ToolResult:
        # prepare() 已经检查过 command，因此这里可以直接读取。
        command = arguments["command"]
        try:
            # 真正启动 PowerShell 的是 CommandRunner 实现，不是这个业务函数。
            result = command_runner.run(str(command), context.workspace)
        except Exception:
            # 不把操作系统的详细异常直接交给模型，避免泄漏本机路径等信息。
            return tool_error("shell_start_failed", "PowerShell process could not be started")

        # 即使命令没有任何输出，也返回明确文本，避免模型误以为漏掉了结果。
        output = result.output or "(no output)"
        if result.truncated:
            output += "\n[output truncated]"
        if result.timed_out:
            return tool_error("shell_timeout", output)
        if result.exit_code != 0:
            return tool_error("shell_failed", f"PowerShell exited with code {result.exit_code}\n{output}")
        return tool_success(output)

    return ToolDefinition(
        name="shell",
        description="Run a PowerShell command in the current workspace.",
        parameters={
            "type": "object",
            "properties": {"command": {"type": "string", "minLength": 1}},
            "required": ["command"],
            "additionalProperties": False,
        },
        effect="execute",
        handler=handler,
    )


def create_chapter_one_tools(command_runner: CommandRunner) -> ToolRegistry:
    """集中注册第 1 章允许模型使用的所有工具。目前只有 shell。"""
    registry = ToolRegistry()
    registry.register(create_shell_tool(command_runner))
    return registry

File: typescript/chapters/ch01/src/features/builtin-tools.ts
import { z } from "zod";

import type { CommandResult, CommandRunner } from "../core/commands.js";
import type { ToolDefinition } from "../core/tools.js";
import { ToolRegistry, toolError, toolSuccess } from "../core/tools.js";

// 第一章内置工具：当前只注册一个受 schema 约束的 shell（PowerShell）工具。
//
// shell 的输入、handler 和副作用类别来自同一个 ToolDefinition，
// 避免模型看到的 schema 与实际执行逻辑平行维护。错误被分类为启动失败、超时和非零退出码，
// 让模型既能读到可操作文本，也能按 errorCode 区分失败原因。
//
// 严格对象拒绝额外字段，命令字符串不得为空；实际命令语义交由 PowerShell。
const shellInputSchema = z.strictObject({ command: z.string().min(1) });

// 第 1 章仅暴露受 schema 约束的 PowerShell 工具，工作目录来自受控上下文。
// commandRunner 是唯一进程边界，工具定义本身不直接启动子进程。
export function createShellTool(commandRunner: CommandRunner): ToolDefinition<{ command: string }> {
  return {
    name: "shell",
    description: "Run a PowerShell command in the current workspace.",
    inputSchema: shellInputSchema,
    effect: "execute",
    handler: async ({ command }, context) => {
      let result: CommandResult;
      try {
        result = await commandRunner.run(command, context.workspace);
      } catch {
        return toolError("shell_start_failed", "PowerShell process could not be started");
      }

      // 结果文本兼容空输出、截断、超时和非零退出码，均可作为下一轮模型上下文。
      let output = result.output.length === 0 ? "(no output)" : result.output;
      if (result.truncated) {
        output = `${output}\n[output truncated]`;
      }
      if (result.timedOut) {
        return toolError("shell_timeout", output);
      }
      if (result.exitCode !== 0) {
        return toolError(
          "shell_failed",
          `PowerShell exited with code ${result.exitCode}\n${output}`,
        );
      }
      return toolSuccess(output);
    },
  };
}

// 创建本章完整工具注册表；新增工具必须在这里集中注册，避免组合根遗漏能力。
export function createChapterOneTools(commandRunner: CommandRunner): ToolRegistry {
  // 章节工具集中注册全部可用副作用，供组装层一次性注入。
  const registry = new ToolRegistry();
  registry.register(createShellTool(commandRunner));
  return registry;
}
